solvers/DeepCORAL.py

```diff
@@ -0,0 +1,125 @@
+    from benchmark_utils.base_solver import DASolver
+    from skada.deep import DeepCoral
+    from skada.deep.base import DomainAwareModule
+    import torch
+    from torch import nn
+    import torchvision.models as models
+    from torch.optim import Adam, Adadelta
+    import collections
+    from skada.metrics import SupervisedScorer, DeepEmbeddedValidation
+
+
+# class ResNet50WithMLP(nn.Module):
+#     def __init__(self, n_classes=2, hidden_size=256, input_shape=(3, 224, 224)):
+#         super().__init__()
+        
+#         self.input_shape = input_shape
+#         in_channels = input_shape[0]  # First dimension is always the number of channels
+
+#         # Load pre-trained ResNet50
+#         resnet = models.resnet50(weights=models.ResNet50_Weights.IMAGENET1K_V1)
+#         # Modify the first layer if input channels != 3
+#         if in_channels != 3:
+#             resnet.conv1 = nn.Conv2d(in_channels, 64, kernel_size=7, stride=2, padding=3, bias=False)
+#         # Remove the last fully connected layer
+#         self.features = nn.Sequential(*list(resnet.children())[:-1])
+#         # To flatten the tensor
+#         self.flatten = nn.Flatten()
+#         # Add 2-layer MLP
+#         self.mlp = nn.Sequential(
+#             nn.Linear(2048, hidden_size),
+#             nn.ReLU(),
+#         )
+#         self.last_layers = nn.Sequential(
+#             nn.Linear(hidden_size, n_classes),
+#             nn.Softmax(dim=1),
+#         )
+
+#     def forward(self, x, sample_weight=None):
+#         x = x.reshape((x.shape[0], *self.input_shape))
+#         x = self.features(x)
+#         x = self.flatten(x)
+#         x = self.mlp(x)
+#         x = self.last_layers(x)
+
+#         return x
+
+
+class ShallowConvNet(nn.Module):
+    # This is a simple convolutional neural network (CNN) for MNIST dataset
+    # https://github.com/pytorch/examples/blob/main/mnist/main.py
+    def __init__(self, n_classes=10, hidden_size=128, input_shape=(1, 28, 28)):
+        super().__init__()
+        
+        self.input_shape = input_shape
+        in_channels = input_shape[0]  # First dimension is always the number of channels
+
+        # Define the layers as per the Net architecture
+        self.conv1 = nn.Conv2d(in_channels, 32, 3, 1)
+        self.conv2 = nn.Conv2d(32, 64, 3, 1)
+        self.dropout1 = nn.Dropout(0.25)
+        self.dropout2 = nn.Dropout(0.5)
+        self.mlp = nn.Linear(9216, hidden_size)
+        self.fc2 = nn.Linear(hidden_size, n_classes)
+
+    def forward(self, x, sample_weight=None):
+        x = x.reshape((x.shape[0], *self.input_shape))
+        x = self.conv1(x)
+        x = nn.ReLU()(x)
+        x = self.conv2(x)
+        x = nn.ReLU()(x)
+        x = nn.MaxPool2d(2)(x)
+        x = self.dropout1(x)
+        x = torch.flatten(x, 1)
+        x = self.mlp(x)
+        x = nn.ReLU()(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = nn.LogSoftmax(dim=1)(x)
+
+        return output
```

Comment:
you should avoid being dependent on the input_shape. You can try something like this:

```Python
import torch
import torch.nn as nn
import torch.nn.functional as F

class ShallowConvNet(nn.Module):
    def __init__(self, n_classes):
        super().__init__()

        # Convolutional layers and first dropout in a ModuleList
        self.convs = nn.ModuleList([
            nn.Conv2d(3, 32, 3, 1),  # First convolutional layer
            nn.ReLU(),
            nn.Conv2d(32, 64, 3, 1),  # Second convolutional layer
            nn.ReLU(),
            nn.Dropout(0.25)  # First dropout layer
        ])
        
        # Global Average Pooling (this layer is fixed and doesn't depend on input size)
        self.global_pool = nn.AdaptiveAvgPool2d((1, 1))
        
        # MLP layers with ReLU and Dropout
        self.mlp = nn.ModuleList([
            nn.Linear(64, 128),  # First MLP layer
            nn.ReLU(),
            nn.Dropout(0.5),  # Dropout after the first MLP layer
            nn.Linear(128, n_classes)  # Output layer directly in MLP
        ])

    def forward(self, x, y=None):
        # Forward pass through convolutional layers and first dropout
        for layer in self.convs:
            x = layer(x)
        
        # Apply global average pooling
        x = self.global_pool(x)
        
        # Flatten the tensor
        x = torch.flatten(x, 1)
        
        # Forward pass through MLP layers
        for layer in self.mlp:
            x = layer(x)
        
        logits = x
        loss = None
        
        if y is not None:
            loss = F.cross_entropy(logits, y)
        
        return logits, loss
```

Reply:
and when data are grayscale (e.g. MNIST) you repeat 3 times the grayscale to mimic RGB

Reply:
I also added `(x, y)` in the forward so that you directly get the loss fct. Not sure it fits well the rest of the code though.

Reply:
Why do you think we need to compute the loss here knowing that its going to be computed again later on ?

Reply:
Yes I think no need to have the loss in the model

Reply:
then your training loop is easy:
```python
optimizer = torch.optim.AdamW(model.parameters(), lr=3e-4)
for i in range(N_EPOCHS):
    optimizer.zero_grad()
    logits, loss = model(x, y)
    print(f"loss: {loss.item()}")
    loss.backward()
    optimizer.step()
```

Reply:
it's small remark, my main point was more about the `input_shape`